### PR TITLE
Refactor txapiclient to add the following:

### PR DESCRIPTION
**Pull request by joshw**
- support for identity API v3
  - uses newer twisted web client (persistent connections, proxy support)
  - simplified client classes with most functionality in base class and
    session manager

Fixes ZPS-938

### CODE REVIEW

File: ZenPacks/zenoss/OpenStackInfrastructure/apiclients/session.py [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+
+        # If the request resulted in an error, raise an exception
+        self.handle_error_response(response, body)
+
+        returnValue((body, response.headers))
+
+    @inlineCallbacks
+    def authenticated_POST_request(self, url, headers=None, body=None):
+        if self.token_id is None:
+            yield self.authenticate()
```
Comment (eimajenthat) [inline ZenPacks/zenoss/OpenStackInfrastructure/apiclients/session.py]: I notice in `authenticated_GET_request`, we have to put the token in the headers.  Why don't we need that for POST?  Other than that, the logic is nearly identical.  Would it make sense to have an authentication wrapper we call for both?

If this is supposed to be different, maybe we should keep them separate.  On the other hand, if it's just something that got added to one, but not the other, it's an example of the dangers of duplicating code.

File: ZenPacks/zenoss/OpenStackInfrastructure/apiclients/txapiclient.py [hunk 2]
```diff
@@ -11,641 +13,152 @@
...
-            for query in kwargs['queries']:
-                self.path += '?q.field=%s&q.op=%s&q.type=%s&q.value=%s' % \
-                        (query.get('field',''), query.get('op',''),
-                         query.get('type',''), query.get('value',''))
+class NovaClient(BaseClient):
+    keystone_service_type = 'compute'
 
-        return self.client.api_call(
-            self.path.encode('ascii', 'ignore'), data=data, params=params, **kwargs)
+    avzones = api('/os-availability-zone/detail')
```
Comment (eimajenthat) [inline ZenPacks/zenoss/OpenStackInfrastructure/apiclients/txapiclient.py]: They really are quite elegant.

File: ZenPacks/zenoss/OpenStackInfrastructure/apiclients/session.py (hunk 1 above)
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/apiclients/session.py]: No, you're right.  We're not actually using authenticated_POST_request anywhere, so I didn't catch the error.

File: ZenPacks/zenoss/OpenStackInfrastructure/apiclients/session.py [hunk 3]
```diff
@@ -0,0 +1,516 @@
...
+
+        try:
+            result = yield self.POST_request(url, headers=headers, body=body)
+            returnValue(result)
+        except UnauthorizedError:
+            log.info("Received unauthorized response to http request, re-authenticating.  (%s)", url)
+            yield self.authenticate()
```
Comment (joshw) [inline ZenPacks/zenoss/OpenStackInfrastructure/apiclients/session.py]: The idea is that the only real situation where we're going to get an unauthorized error after having successfully authenticated is that our token has expired- so it should only take one attempt to get a new one and be good to go.  More than that, and something's more broken, and recovery isn't likely, so i'd rather just let it fail.

Comment (eimajenthat) [inline ZenPacks/zenoss/OpenStackInfrastructure/apiclients/session.py]: Ah, that makes sense.

**Comment by eimajenthat** [comment]
@joshw Approved.

**Pull request merged**